### ISSUE TITLE
Triclinic Distance Bug

### ISSUE THREAD

**Issue by rmcgibbo** [mdtraj/mdtraj#1019]
I found another bug in `md.compute_distances` with triclinic unit cells. Unfortunately, I don't think #930 was quite correct in the general case.

```
import numpy as np
import mdtraj as md
from scipy.spatial.distance import euclidean

t = md.Trajectory(
    xyz=np.array([[0.5338, 1.5336, 0.9745], [1.4097, 1.9486, 1.1537]]),
    unitcell_lengths=np.array([[2., 2., 2.]]),
    unitcell_angles=np.array([[46.8, 34.4, 78.7]]),
    topology=None)


def simple_brute_force(v1, v2, a, b, c):
    d = float('infinity')
    for i in range(-2, 2):
        for j in range(-2, 2):
            for k in range(-2, 2):
                v2_p = v2 + (i * a + j * b + k * c)
                d = min(d, euclidean(v1, v2_p))
    return d


print('md.compute_distances (opt=True)', md.compute_distances(t, [[0, 1]], opt=True))
print('md.compute_distances (opt=False)', md.compute_distances(t, [[0, 1]], opt=False))
print('simple_brute_force',
      simple_brute_force(v1=t.xyz[0, 0],
                         v2=t.xyz[0, 1],
                         a=t.unitcell_vectors[0, 0],
                         b=t.unitcell_vectors[0, 1],
                         c=t.unitcell_vectors[0, 2]))
```

```
md.compute_distances (opt=True) [[ 0.74414033]]
md.compute_distances (opt=False) [[ 0.74414021]]
simple_brute_force 0.6186748147010803
```

**Comment by swails** [comment]
This isn't an orthorhombic unit cell...

**Comment by rmcgibbo** [comment]
Sorry, I got those words confused. I meant triclinic.

**Comment by rmcgibbo** [comment]
I mean the complicated one.

**Comment by swails** [comment]
Well one difference between your code and the code in geometry is that you loop over displacements -2, -1, 0, and 1.  Whereas [the code in mdtraj right now](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/geometry/distance.py#L239-L248) only loops over -1, 0, and 1.  Basically, the idea is that the "standard" MIC code (which works for MD because the distances it gets *wrong* must be too far away to lie within the maximum permissible cutoff) gets the particle *close enough* that you only have to test images in the 27 adjacent unit cells.

Your code is looking over 64 adjacent unit cells...  Clearly my assumption wasn't quite accurate.  I guess for *extremely* long and skinny unit cells, you may have to look farther away than the 27 adjacent images?  So you can try changing the `range(-2, 1)` to `range(-2, 3)` in the lines I linked and see if that fixes the non-optimized version.  If it does, then I'm afraid I really don't know the way to fix this rigorously without introducing absurd overhead.  I'm not sure how to algorithmically determine the maximum number of images you'd need to search to be sure of getting the right answer based on the unit cell dimensions.  For "standard" ones (i.e., dodecahedron, octahedron, and cubic), the current code I think works correctly.

**Comment by swails** [comment]
I mean changing `range(-1, 2)` to `range(-2, 3)`

**Comment by rmcgibbo** [comment]
I already have a fix for this, yes, but I found another. In fact, looping over 2 is not sufficient in the general case.

**Comment by swails** [comment]
I took this general idea from what cpptraj does, which is just looking at the 26 adjacent unit cells...

> I already have a fix for this, yes, but I found another. In fact, looping over 2 is not sufficient in the general case.

Well I suspect that the more acute you make the angles, the wider your range of unit cell vectors will need to be.  Is that what you found?

**Comment by rmcgibbo** [comment]
The code from #930 looks over the adjacent cells, but actually does it in a different way than cpptraj does, because it first calculates a displacement vector in real coordinates and then converts that to the fractional coordinate system and tries to normalize it, whereas cpptraj converts each position to fractional coordinates, and then takes the difference. This is before searching through the 26 displacements to the displacement vector.

I found this bug by fuzzing the mdtraj code against cpptraj to find where they differ.

But then I started fuzzing the cpptraj code against the super-brute-force version, and found https://github.com/Amber-MD/cpptraj/issues/223...

**Comment by rmcgibbo** [comment]
Honestly, this should probably just go with #1015 as wishlist.

**Comment by rmcgibbo** [comment]
No, stays open. It's still a bug.

**Comment by mpharrigan** [comment]
My b

**Comment by mpharrigan** [comment]
Fixed as much as this is going to be fixed in #1077
